### ISSUE TITLE
[Feature]: VolumeSnapshot retention policy

### ISSUE THREAD

**Issue by nhudson** [cloudnative-pg/cloudnative-pg#6009]
### Is there an existing issue already for this feature request/idea?

- [x] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

Currently in CNPG when you have volume snapshots enabled the `.spec.backup.retentionPolicy` is ignored.  This leads to having to implement your own retention policy or manually delete older snapshots as they grow unbounded in the cluster over time.

### Describe the solution you'd like

Add functionality to remove snapshots older than what is set in the `.spec.backup.rententionPolicy` in the same manner as what is being done with Barman backups.

When we have snapshots older than `.spec.backup.retentionPolicy` is set to, delete snapshots for the specific instance.

### Describe alternatives you've considered

We currently are in the process of adding some logic into a customer controller to watch and handle CNPG backups that are older than what is set in `.spec.backup.rententionPolicy` for each deployed `Cluster`.  This would basically look at all `Backups` and if any are older than the retention policy value then it would delete the `Backup` and ensure that the `VolumeSnapshot` and `VolumeSnapshotContents` are deleted from the cluster.

### Backport?

N/A

### Are you willing to actively contribute to this feature?

Yes

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by lennartack** [comment]
I think CNPG should implement this rather than something else. VolumeSnapshots are generic snapshots that may be used for many different purposes, so you typically don't want a cluster-wide retention policy for VolumeSnaphots. Rather, you'd want a retention policy for ScheduledBackups/Backups, which are custom resources from CNPG. It only makes sense that CNPG implements retention for these.

**Comment by rafal11ck** [comment]
Still issue, at least docs say so.

**Comment by joaopedrocg27** [comment]
Any updates on this?

**Comment by horihel** [comment]
If there's different retention policies for volumeSnapshot and barman-plugin, I suspect there's a possibility that a user could end up with a base backup that can't be rolled forward from the WAL archive. If both basebackup and WAL are managed by barman, then barman takes care of this dependency and will make sure to keep all base backups and archives that are required to fulfil the recovery window. But barman doesn't know about volumeSnapshots and cnpg doesn't know how long barman keeps WAL archives.

cnpg doesn't know barman's retention policy - maybe it makes even more sense to have barman-plugin handle volumeSnapshots (or at least allow it to know or share the retention policy?)

**Comment by d-a-n-k-a** [comment]
it's all mowing towards [CNPG-I](https://github.com/cloudnative-pg/cnpg-i) 👍 ,

[Retention Policies](https://github.com/cloudnative-pg/cloudnative-pg/blob/main/docs/src/backup.md#retention-policies)

> CloudNativePG is evolving toward a backup-agnostic architecture, where backup responsibilities are delegated to external CNPG-I plugins. These plugins are expected to offer advanced and customizable data protection features, including sophisticated retention management, that go beyond the built-in capabilities and scope of CloudNativePG.
>
> As part of this transition, the spec.backup.**retentionPolicy field in the Cluster resource is deprecated** and will be removed in a future release.

**Comment by d-a-n-k-a** [comment]
@gbartolini , I think this issue can be closed in favour of #7341

**Comment by horihel** [comment]
Retention and the resulting restore window is data that has to be known and shared between both mechanism (volumesnapshots and barman). Not sure if cnpg-i allows cooperation/data-sharing between plugins. Can barman-cloud invoke snapshot restores? Or the other way around - could a volsnap-plugin invoke the barman plugin to replay the WAL?

**Comment by d-a-n-k-a** [comment]
I suppose focusing in cold backups only -- no need for Barman and S3 in this case.

**Comment by horihel** [comment]
see https://cloudnative-pg.io/documentation/1.26/backup/#cold-and-hot-backups
>Native volume snapshots:
>
>   Do not require WAL archiving, though its use is still strongly recommended in production

if it's "strongly recommended for production", then any implementation that doesn't allow for it, wouldn't be fit for production.

**Comment by dragoangel** [comment]
No stale please, this is really important to have ability to automatically rotate backups created by specific schedule. I think it should work under backup crd finilizer logic: deletion of backup object should trigger deletion files in S3 or volumeSnapshot, if it fails - deletion of backup also fails, if it success - backup resource will be deleted. 

I believe that it's k8s administrator responsibility to have retainPolicy for snapshot/pv to be set to delete nowadays, if they deploying backup rotation and not understand this it's not an Cloudnative fault's.

**Comment by rouke-broersma** [comment]
@d-a-n-k-a @gbartolini since volume snapshot based backups are not moving to cnpg-i, this issue has become relevant again. CNPG currently does not implement any retention policy for volume snapshot based backups, causing unbounded growth unless the user creates their own snapshot cleanup mechanism.

**Comment by NiklasBeierl** [comment]
I wonder if @gbartolini can already lay out what the pitfalls with retention for VolumeSnapshots will be. The closing comment on #7341 gives me the impression that there was already quite some thought and discussion on the matter?

From how I intend to use it (running cnpg on top of longhorn) it should actually be quite simple, deleting Backups would trigger the deletion of VolumeSnapshots (ownerReferences), which in turn causes longhorn to merge the snapshots. From what I can tell, all that would be needed in this setup is checking the number / age of backups and deleting the "oldest" ones according to policy. I'm almost tempted to just implement that for myself. I don't really know if it is that simple for all storage drivers, tho.

**Comment by dragoangel** [comment]
@NiklasBeierl you not care about csi plugin person used to implement that, same as you not care how to create backup on different storages, all you say is: `VolumeSnapshotClass` name in `snapshot.storage.k8s.io/v1` to call. So you care only about finalizer of backup where method is set to `VolumeSnapshot`: drop referenced `VolumeSnapshot` from kube API. Person who will expect retention policy to work must configure his snapshot class to drop snapshots from storage when CR of snapshot dropped.

**Comment by dragoangel** [comment]
@NiklasBeierl want also to note one small important part: [Persistence of volume snapshot objects](https://cloudnative-pg.io/documentation/current/appendixes/backup_volumesnapshot/#persistence-of-volume-snapshot-objects) states:

> By default, VolumeSnapshot objects created by CloudNativePG are retained after deleting the Backup object that originated them, or the Cluster they refer to.

if you set `.spec.backup.volumeSnapshot.snapshotOwnerReference` to `backup` and delete `Backup` CR it should already trigger deletion of related `VolumeSnapshot`

**Comment by NiklasBeierl** [comment]
@dragoangel Yup, all of what you wrote aligns with my understanding of the Snapshot API. 

I am just wondering if there is any need for the retention mechanism in CNPG to be any more involved than: "Delete the Backup CR after X time" / "Delete the Backup CR if there is N newer ones". 

I guess not? As I understand, VolumeSnapshots should be crash consistent, so I can use any snapshot to go back in time. Essentially I am pretending that someone pulled the plug on my DB server at the time when the Snapshot was created. So every VolumeSnapshot is a complete / full backup.  They can be used for restore / recovery independently of one another and thus their retention can also be managed independently.

**Comment by dragoangel** [comment]
Volume snapshots not depends on your pgsql server, they depends on object storage which reports about fact if snapshot ok or not. You can drop your cluster and still snap your volume if it persisted.

What I think off is that scheduled backup must have retention policy, and operator need to understand what backup was issued by what schedule and rotate it accordingly

**Comment by dragoangel** [comment]
Unstale please

**Comment by lambcode-unified** [comment]
Unstale please. This is an important missing feature in cnpg.

**Comment by ncgee** [comment]
It would be nice to have a solution for this. We're working around it by adding a `kube-janitor` annotation to the backups via the cluster's `spec.backup.volumeSnapshot.annotations`, but this only adds the annotation to the VolumeSnapshot itself, not the `Backup` object, so we now have backup objects whose associated volume snapshots have been deleted.

I'm kind of confused by the lack of retention configuration in the ScheduledBackup specification, it does not make much sense to me to associate it with the object store configuration. For example, if I want hourly backups and keep them for a day, and daily backups for let's say 4 weeks, it would make sense to me to set both retention policies with their respective scheduled backups.

**Comment by dragoangel** [comment]
@ncgee discussed above

> > By default, VolumeSnapshot objects created by CloudNativePG are retained after deleting the Backup object that originated them, or the Cluster they refer to.
> 
> if you set `.spec.backup.volumeSnapshot.snapshotOwnerReference` to `backup` and delete `Backup` CR it should already trigger deletion of related `VolumeSnapshot`


So technically temp workaround is `kube-janitor` + `spec.backup.volumeSnapshot.annotations` + `spec.backup.volumeSnapshot.snapshotOwnerReference=backup` - possible, but far not ideal due to overcomplexity.

**Comment by lambcode-unified** [comment]
Please unstale. We are waiting for this feature too.
